File: program/deal_ice_dict_to_xh.js
const fs = require('fs');
const path = require('path');

// dict文件的路径
const dictFilePath = path.join(__dirname, '../flypy_flypy.dict.yaml');

// 同步读取YAML文件
const dictFileContent = fs.readFileSync(dictFilePath, 'utf8');

// 获取每个字的编码
//阿	aaek
// 按行分割文本数据
const dictLines = dictFileContent.split('\n');

// 解析每一行数据
const dictData = {};
dictLines.forEach((line) => {
    // 假设每行的格式为 "字\t编码"
    if (line.indexOf("\t") != -1) {
        const [character, encoding] = line.split('\t');
        if (encoding.indexOf("'") == -1) {
            var encoding_pre = encoding.substring(0, 2);
            var encoding_post = encoding.substring(3);
            dictData[character] = encoding_post;
        }
    }
});

// 获取每个字的编码
// console.log('编码:', dictData);
console.log("孙 " + dictData['孙'])

// 需要修改的YAML文件的路径
const yamlFilePath = path.join(__dirname, '../cn_dicts/base.dict.yaml');

// 同步读取YAML文件
const yamlFileContent = fs.readFileSync(yamlFilePath, 'utf8');

// 按行分割文本数据
const yamlLines = yamlFileContent.split('\n');


// Function to read a file
function readFile(filePath) {
    return fs.readFileSync(filePath, 'utf8');
}

// Function to write content to a file
function writeFile(filePath, content) {
    fs.writeFileSync(filePath, content, 'utf8');
}

// Function to update missing encodings in the file
function updateMissingEncodings(filePath) {
    // Read the file content
    const fileContent = readFile(filePath);

    // Split the content into lines
    const lines = fileContent.split('\n');

    // Create an updated content variable
    let updatedContent = '';

    // Process each line
    lines.forEach((line) => {
        // Assume each line has the format "字\t编码"
        if (line.includes('\t') && !line.startsWith("#")) {
            const [character, encoding, frequency] = line.split('\t');

            // console.log(encoding)

            pinyin_list = encoding.split(" ")
            double_list = ""
            var pinyin_index = 0
            for (pinyin of pinyin_list) {
                double_pinyin = ""
                if (pinyin.length == 2) {
                    double_pinyin = pinyin
                } else if (pinyin.length == 1) {
                    double_pinyin = pinyin + pinyin
                } else {
                    shengmu = pinyin.substring(0, 1)
                    yunmu = pinyin.substring(1)

                    double_shengmu = shengmu
                    if (pinyin.charAt(1) == "h") {
                        shengmu = pinyin.substring(0, 2)
                        yunmu = pinyin.substring(2)
                        shengmu_map = { "zh": "v", "ch": "i", "sh": "u" }
                        double_shengmu = shengmu_map[shengmu]
                    }
                    if (double_shengmu == 'a') {
                        yunmu = pinyin
                    }
                    if (double_shengmu == 'o') {
                        yunmu = pinyin
                    }
                    const yunmu_map = {
                        'iu': 'q',
                        'ei': 'w',
                        'uan': 'r',
                        'van': 'r',

                        'ue': 't',
                        've': 't',
                        'un': 'y',
                        'vn': 'y',
                        'uo': 'o',
                        'ie': 'p',

                        'iong': 's',
                        'ong': 's',

                        'ai': 'd',
                        'en': 'f',

                        'eng': 'g',
                        'ang': 'h',
                        'an': 'j',

                        'ing': 'k',
                        'uai': 'k',
                        'iang': 'l',
                        'uang': 'l',

                        'ou': 'z',
                        'ia': 'x',
                        'ua': 'x',
                        'ao': 'c',
                        'ui': 'v',
                        'in': 'b',

                        'iao': 'n',

                        'ian': 'm',
                    };
                    double_yummu = yunmu_map[yunmu]
                    if (yunmu.length == 1)
                        double_yummu = yunmu

                    double_pinyin = double_shengmu + double_yummu

                    if (double_pinyin.length != 2) {
                        console.log("!!!!double_pinyin " + double_pinyin + " " + pinyin + " ")
                    }
                }
                var clean_character = character.replace("·", "")
                var character_encoding_pre = clean_character.charAt(pinyin_index)
                encoding_post = dictData[character_encoding_pre];
                if (encoding_post == null) {
                    // console.log("!!!character_encoding_pre:" + character_encoding_pre)
                    // console.log("!!!!encoding_post==null " + double_pinyin + " " + pinyin + " character_encoding_pre:" + character_encoding_pre + " " + line)
                    // console.log("character:" + character)
                    // console.log("pinyin_index " + pinyin_index)
                    // console.log("pinyin " + pinyin)
                    // console.log("character.charAt(pinyin_index):" + character.charAt(pinyin_index))
                    // console.log("double_pinyin " + double_pinyin)
                }
                double_list += double_pinyin + "[" + encoding_post
                double_list += " "
                pinyin_index++
            }
            double_list = double_list.substring(0, double_list.length - 1)
            // Update the line with the missing encoding
            const updatedLine = `${character}\t${double_list}\t${frequency}`;
            console.log(updatedLine)
            // Append the updated line to the updated content
            updatedContent += updatedLine + '\n';
        } else {
            // Append non-matching lines to the updated content
            updatedContent += line + '\n';
        }
    });

    // Write the updated content back to the file
    writeFile(filePath, updatedContent);
}


// Call the function to update missing encodings in the file
updateMissingEncodings(yamlFilePath);

File: program/deal_ice_dict_to_xh.py
import os

# Function to read a file
def read_file(file_path):
    with open(file_path, 'r', encoding='utf-8') as file:
        return file.read()

# Function to write content to a file
def write_file(file_path, content):
    with open(file_path, 'w', encoding='utf-8') as file:
        file.write(content)

# Function to update missing encodings in the file
def update_missing_encodings(file_path, write_file_path, dict_data):
    # Read the file content
    file_content = read_file(file_path)

    # Split the content into lines
    lines = file_content.split('\n')

    # Create an updated content variable
    updated_content = ''

    # Process each line
    for line in lines:
        if '\t' not in line or line.startswith("#"):
            updated_content += line + '\n'
            continue

        frequency = None
         # 检查分割出的值是否足够
        if len(line.split('\t')) == 3:
            character, encoding, frequency = line.split('\t')
        else:
            character, encoding = line.split('\t')
        
        
        if "tencent" in file_path :
            updated_line = f"{character}\t99"
            updated_content += updated_line + '\n'
            continue
        else:
            if encoding == "100":
                updated_content += line + '\n'
                continue

        pinyin_list = encoding.split(" ")
        double_list = ""
        pinyin_index = 0
        for pinyin in pinyin_list:
            if len(pinyin) == 2:
                double_pinyin = pinyin
            elif len(pinyin) == 1:
                double_pinyin = pinyin + pinyin
            else:
                shengmu = pinyin[0]
                yunmu = pinyin[1:]

                double_shengmu = shengmu
                if pinyin[1] == "h":
                    shengmu = pinyin[:2]
                    yunmu = pinyin[2:]
                    shengmu_map = { "zh": "v", "ch": "i", "sh": "u" }
                    double_shengmu = shengmu_map[shengmu]
                
                if double_shengmu == 'a':
                    yunmu = pinyin
                
                if double_shengmu == 'o':
                    yunmu = pinyin
                
                if double_shengmu == 'e':
                    yunmu = pinyin
                
                yunmu_map = {
                    'iu': 'q',
                    'ei': 'w',
                    'uan': 'r',
                    'van': 'r',

                    'ue': 't',
                    've': 't',
                    'un': 'y',
                    'vn': 'y',
                    'uo': 'o',
                    'ie': 'p',

                    'iong': 's',
                    'ong': 's',

                    'ai': 'd',
                    'en': 'f',

                    'eng': 'g',
                    'ang': 'h',
                    'an': 'j',

                    'ing': 'k',
                    'uai': 'k',
                    'iang': 'l',
                    'uang': 'l',

                    'ou': 'z',
                    'ia': 'x',
                    'ua': 'x',
                    'ao': 'c',
                    'ui': 'v',
                    'in': 'b',

                    'iao': 'n',

                    'ian': 'm',
                };
                
                if len(yunmu) == 1:
                    double_yummu = yunmu
                else:
                    double_yummu = yunmu_map[yunmu]
                double_pinyin = double_shengmu + double_yummu

                if len(double_pinyin) != 2:
                    print("!!!!double_pinyin " + double_pinyin + " " + pinyin + " ")
                
            clean_character = character.replace("·", "")
            character_encoding_pre = clean_character[pinyin_index]
            # character_encoding_pre = character[pinyin_index]
            encoding_post = dict_data.get(character_encoding_pre, "[")
            double_list += f"{double_pinyin}[{encoding_post} "
            pinyin_index += 1
        
        double_list = double_list[:-1]

        if "[[" in double_list:
            # updated_content += line + '\n'
            pass

        if "tencent" in file_path :
            updated_line = f"{character}\t99"
        else:
            if frequency is not None:
                updated_line = f"{character}\t{double_list}\t{frequency}"
            else :
                updated_line = f"{character}\t{double_list}"
        updated_content += updated_line + '\n'

    # Write the updated content back to the file
    write_file(write_file_path, updated_content)

dict_data = {}
file_list = ['8105.dict.yaml', '41448.dict.yaml', 'base.dict.yaml', 'ext.dict.yaml', 'others.dict.yaml']
# file_list = [ 'tencent.dict.yaml']
# Load the dict data from the provided file
with open('./program/flypydz.yaml', 'r', encoding='utf-8') as dict_file:
    for line in dict_file:
        if "\t" in line:
            character, encoding = line.strip().split('\t')
            if "'" not in encoding:
                encoding_pre = encoding[:2]
                encoding_post = encoding[2:]
                # if character in '去我而人他有是出哦配啊算的非个和就可了在小从这吧你吗':
                #     if len(encoding_post) == 2:
                #         encoding_post = encoding_post[0] + encoding_post[1].upper()
                if character not in dict_data:
                    dict_data[character] = encoding_post


print("巴 " + dict_data['巴'])
print("𬱖 " + dict_data['𬱖'])

for file_name in file_list:
    # File paths
    cn_dicts_path = os.path.expanduser("~/vscode/rime-frost/cn_dicts")
    yaml_file_path = os.path.join(cn_dicts_path, file_name)
    write_file_path = os.path.join('cn_dicts_xh', file_name)

    print(yaml_file_path)
    # Update missing encodings in the file
    update_missing_encodings(yaml_file_path, write_file_path, dict_data)


yaml_file_path = os.path.join('cn_dicts_dazhu', 'cell.dict.yaml')
write_file_path = os.path.join('cn_dicts_xh', 'cell.dict.yaml')

print(yaml_file_path)
# Update missing encodings in the file
update_missing_encodings(yaml_file_path, write_file_path, dict_data)
